### PR TITLE
fix: Service restart handling with watch requisite

### PR DESCRIPTION
**Pull request by danrodrig**
### PR progress checklist (to be filled in by reviewers)

- [x] Changes to documentation are appropriate (or tick if not required)
- [x] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Describe the changes you're proposing

Previously, the state for the keepalived service included the `enable`
attribute. This would prevent the watch requisite to restart the service
if the service was already running but not enabled.

Split service handling in `service.enabled` and `service.running`.

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### Additional context

For reference, https://docs.saltproject.io/en/latest/ref/states/requisites.html#watch

### CODE REVIEW

**Comment by danrodrig** [comment]
@myii Could you take a look?

**Comment by myii** [comment]
@danrodrig Apologies for the delayed response.

It's surprising that you're having to resort to the `service.enabled` state -- there are virtually no other examples of this [across the formulas in this org](https://github.com/search?l=SaltStack&q=org%3Asaltstack-formulas+%22service.enabled%22&type=Code).

The [documentation](https://docs.saltproject.io/en/master/ref/states/all/salt.states.service.html#salt.states.service.enabled) for the state also suggests using `enable: True`.

I decided to run a test anyway and specifically disabled the state while it was running, before rerunning the formula -- the service was enabled without issues.

Perhaps you could share your setup (`salt -V`), to figure out where this issue is arising from.

**Comment by danrodrig** [comment]
> @danrodrig Apologies for the delayed response.
> 
> It's surprising that you're having to resort to the `service.enabled` state -- there are virtually no other examples of this [across the formulas in this org](https://github.com/search?l=SaltStack&q=org%3Asaltstack-formulas+%22service.enabled%22&type=Code).
> 
> The [documentation](https://docs.saltproject.io/en/master/ref/states/all/salt.states.service.html#salt.states.service.enabled) for the state also suggests using `enable: True`.

It's mentioned in the documentation for the watch requisite,

If your service is already running but not enabled, you might expect that Salt will be able to tell that since the config file changed your service needs to be restarted. This is not the case. Because the service needs to be enabled, that change will be made and mod_watch will never be triggered.

Splitting the states is the suggested approach,

Now that service.running is its own state, changes to service.enabled will no longer prevent mod_watch from getting triggered, so your httpd service will get restarted like you want.

> I decided to run a test anyway and specifically disabled the state while it was running, before rerunning the formula -- the service was enabled without issues.
> 

To test this, the service must be disabled and running, and a change in `sls: {{ sls_config_file }}` needs to trigger the watch.

> Perhaps you could share your setup (`salt -V`), to figure out where this issue is arising from.

I can prepare something if you need it.

**Comment by hatifnatt** [comment]
@myii problem is not with enabling the service, problem with service restart when required by `watch` condition and when service is running but it is not enabled.
Check https://github.com/saltstack/salt/issues/47361 for details. Usually this issue is fixed by triggering `service.restart` module
like here https://github.com/saltstack-formulas/zabbix-formula/pull/77/files

@danrodrig so, separation of `enable: true` to dedicated state helps with correctly restarting disabled service by `watch` requisite? I think it can be viable solution. And it's better then using `module.wait` due changes in `module.*` state syntax.

**Comment by myii** [comment]
> To test this, the service must be disabled and running, and a change in `sls: {{ sls_config_file }}` needs to trigger the watch.
> 
> > Perhaps you could share your setup (`salt -V`), to figure out where this issue is arising from.
> 
> I can prepare something if you need it.

@danrodrig I've even tried this with the `default-debian-10-master-py3` instance but can't reproduce.  That's why I'm asking for the `salt -V` output for both master and minion.

---

> Check [saltstack/salt#47361](https://github.com/saltstack/salt/issues/47361) for details.

@hatifnatt Thanks for sharing that.  If I get a moment, I'll try running one of those examples to see if I can reproduce.  Also, do you have a specific platform/salt version that you can reproduce the issue with?

**Comment by hatifnatt** [comment]
> Also, do you have a specific platform/salt version that you can reproduce the issue with?

This issue is present since v2018 (and it was present in earlier versions I think) and due fact that https://github.com/saltstack/salt/issues/47361 is still open I think it is not fixed yet.

People usually does not face this issue because that's a rare situation when service disabled, and it's running and restart is required by `watch` requisite.  I.e. if service is disabled and it's not running - there is no problem at all, service will be correctly enabled and started.

**Comment by danrodrig** [comment]
> @danrodrig so, separation of `enable: true` to dedicated state helps with correctly restarting disabled service by `watch` requisite? I think it can be viable solution. And it's better then using `module.wait` due changes in `module.*` state syntax.

@hatifnatt this problem and the solution is already in the [watch requisite documentation](https://docs.saltproject.io/en/latest/ref/states/requisites.html#watch). Just at the end before the section for `listen`, starts with, `One common source of confusion is expecting mod_watch...`

File: keepalived/service/running.sls [hunk 1]
```diff
@@ -12,6 +12,9 @@ include:
 keepalived-service-running-service-running:
   service.running:
     - name: {{ keepalived.service.name }}
-    - enable: True
     - watch:
       - sls: {{ sls_config_file }}
+
+keepalived-service-enabled:
```
Comment (hatifnatt) [inline keepalived/service/running.sls:18]:
```suggestion
keepalived-service-running-service-enabled:
```
One small change I can suggest - change state id to match with formula naming scheme: `path-to-slsfile-state-name` yes it's look little bit weird, I didn't like it at first, but it helps to track which file state belongs.

**Review by hatifnatt — CHANGES_REQUESTED** [review]

File: keepalived/service/running.sls (hunk 1 above)
Comment (danrodrig) [reply keepalived/service/running.sls]: yes, it makes sense to keep them consistent.

**Review by hatifnatt — APPROVED** [review]

**Review by myii — APPROVED** [review]

**Comment by myii** [comment]
Merged, thanks for the PR @danrodrig.  Appreciate the feedback from both you and @hatifnatt (and the review!).

---

@saltstack-formulas/wg The circumstances of this PR is worth covering in the next working group meeting.  This may be something we should be doing across all formulas, rather than recommending `enable: True` as part of the `service.running` state.  If that is the case, then we should use this in the `template-formula` as well.

**Pull request merged**